url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPLinkManager.java

```diff
@@ -0,0 +1,2057 @@
+        }
+    }
+
+    /**
+     * Deregister all link status handlers.
+     */
+    public final void deregisterAllLinkStatusHandlers() {
+        synchronized (linkStateHandlers) {
+            linkStateHandlers.clear();
+        }
+    }
+
+    /**
+     * Register a handler for a local task ID. This handler will receive all OCP
+     * call control messages for the registered task ID. Only one handler can be
+     * registered at a time for a given task ID - if the task ID is already in
+     * use, the existing handler will be removed.
+     * <p>
+     * Each registered handler counts as an active call.
+     *
+     * @param handler
+     *            The handler to register.
+     * @param tid
+     *            The OCP Task ID that the handler is interested in.
+     * @see #deregisterTidHandler(int)
+     */
+    @Override
+    public final void registerTidHandler(final OCPMessageHandler handler,
+            final int tid) {
+        synchronized (tidHandlers) {
+            tidHandlers.put(tid, handler);
+        }
+    }
+
+    /**
+     * Deregister a handler for a local task ID.
+     *
+     * @param tid
+     *            The task ID to remove the handler for.
+     * @see #registerTidHandler(OCPMessageHandler, int)
+     */
+    @Override
+    public final void deregisterTidHandler(final int tid) {
+        synchronized (tidHandlers) {
+            tidHandlers.remove(tid);
+        }
+    }
+
+    /**
+     * Register a handler for the management task ID. This handler will receive
+     * all OCP call control messages for the task ID 0xFFFFFFFF. Only one
+     * handler can be registered at a time for a given task ID - if the task ID
+     * is already in use, the existing handler will be removed.
+     * <p>
+     * This handler does not count as an active call.
+     *
+     * @param handler
+     *            The handler to register.
+     * @see #deregisterManagementTidHandler()
+     */
+    public final void registerManagementTidHandler(
+            final OCPMessageHandler handler)    {
+        managementTidHandler = handler;
+    }
+
+    /**
+     * Deregister a handler for the management task ID.
+     *
+     * @see #registerManagementTidHandler(OCPMessageHandler)
+     */
+    public final void deregisterManagementTidHandler() {
+        managementTidHandler = null;
+    }
+
+    /**
+     * Queue an OCP message for transmitting to the remote unit.
+     *
+     * @param message
+     *            The OCP message to be processed by the implementation.
+     * @param callingLink
+     *            Not used.
+     */
+    @Override
+    public final void queueMessage(final OCPMessage message,
+            final OCPLink callingLink) {
+        // no need to synchronise as BlockingQueues are thread-safe
+        try {
+            stats.incrementStat(statPrefix + STAT_QUEUE);
+            txMessages.put((LegacyOCPMessage) message);
+        } catch (InterruptedException e) { // CSIGNORE: EmptyBlock
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)